src/mcts/search.cc

```diff
@@ -907,6 +973,19 @@ SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend(
         ++possible_moves;
       }
       float Q = child.GetQ(fpu);
+
+      // Certainty Propagation. Avoid suboptimal childs.
+      if (params_.GetCertaintyPropagation()) {
+        // Prefers lower bounded childs over drawing children.
+        if (child.edge()->IsOnlyLBounded() && child.GetQ(0) <= 0.0f) Q = 0.01f;
+        // Prefers drawing children over upper bounded childs.
+        if (child.edge()->IsOnlyUBounded() && child.GetQ(0) >= 0.0f) Q = -0.01f;
+        // Penalize exploring suboptimal childs throughout the tree.
+        if (parent_upperbounded) {
+          if (child.edge()->IsOnlyUBounded()) Q -= child.GetN() * 0.1f;
```

Comment:
This could be a massive subtraction - again 0.1 seems a magic number.  Should the subtraction really be able to grow unbounded?

Reply:
Yes, it can grow unbounded which is no problem at all and desired behaviour (at least its supposed to be) . Let me first explain what this does:
(1) if the parent is upperbounded, we know that the player to move is lowerbounded implying that we already have at least a draw if not more. This further implies that we must have at least one edge (move) that is lower-bounded or else the parent would not have been upper-bounded.
(2) With this knowledge we do not need to search any upper-bounded moves at all (!), as these give at maximum a draw - remember we know that we got at least one move that is at least a draw.  All visits should be allocated to the non-upper bounded moves. So why not just hard prune them and be done with it. Well....
(3) There is a performance trade-off of "certain" (or terminal) nodes vs. NN evals. In my testing chosing these suboptimals (but bounded)  lines often quickly runs into certain nodes or terminal nodes yielding results faster than fetching real NN evals going deeper in the optimal line. 
(3) The scores of these visits will be adjusted when flowing through the bounded parent node, so the difference in scores propagated upward between taking these optimal visits or the suboptimal ones initially is minor.
(4) So the above code does not hard prune these nodes, but visits them with decreasing likelihood, so that if initially the probability is high hitting (fast) terminals/certain nodes we visit those just like "normal" leela would. If then leela stubbornly wants to continue visiting these branches, despite being suboptimal we (more or less gradually) increase the penality for doing so.

Now as for the size of the penality, this is more of a magic number than the other epsilon. I can't really say i did much testing on this. I basically eyeballed this that after 10 visits i really do not want to visit this suboptimal line again. Short version: Growing unbounded is no problem, but number can probably be optimized.